src/Features/CSharp/Portable/CodeRefactorings/ConvertLocalFunctionToMethod/CSharpConvertLocalFunctionToMethodCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,230 @@
+
+                var updatedInvocation = currentInvocation.AddArgumentListArguments(capturesAsArguments.ToArray());
+
+                if (enclosingTypeParameters != null)
+                {
+                    var typeArguments = enclosingTypeParameters.Concat(symbol.TypeArguments.Select(x => x.GenerateTypeSyntax()));
+                    updatedInvocation = updatedInvocation.WithExpression(
+                            (ExpressionSyntax)generator.WithTypeArguments(updatedInvocation.Expression, typeArguments));
+                }
+
+                editor.ReplaceNode(currentInvocation, updatedInvocation);
+            }
+
+            // Update names only if needed
+            if (methodName != declaredSymbol.Name)
+            {
+                var nameToken = methodName.ToIdentifierToken();
+                foreach (var id in parentBlock.DescendantNodes().OfType<IdentifierNameSyntax>())
+                {
+                    var symbol = semanticModel.GetSymbolInfo(id, cancellationToken).Symbol as IMethodSymbol;
+                    if (symbol?.OriginalDefinition == declaredSymbol)
+                    {
+                        editor.ReplaceNode(id, WithName(id, nameToken));
+                    }
+                }
+            }
+
+            return document.WithSyntaxRoot(editor.GetChangedRoot());
+        }
+
+        private static ExpressionSyntax WithName(ExpressionSyntax expression, SyntaxToken name)
+            => expression is SimpleNameSyntax n ? n.WithIdentifier(name) : expression;
+
+        private static bool SupportsNonTrailingNamedArguments(ParseOptions options)
+            => ((CSharpParseOptions)options).LanguageVersion >= LanguageVersion.CSharp7_2;
+
+        private static void GetCapturedTypeParameters(ISymbol symbol, List<ITypeParameterSymbol> typeParameters)
+        {
+            var containingSymbol = symbol.ContainingSymbol;
+            if (containingSymbol != null && containingSymbol.Kind != SymbolKind.NamedType)
+            {
+                GetCapturedTypeParameters(containingSymbol, typeParameters);
+            }
+
+            typeParameters.AddRange(symbol.GetTypeParameters());
+        }
+
+        private static void RemoveUnusedTypeParameters(
+            SyntaxNode localFunction,
+            SemanticModel semanticModel,
+            List<ITypeParameterSymbol> captureTypes,
+            List<ITypeParameterSymbol> typeParameters)
+        {
+            var unusedTypeParameters = typeParameters.ToList();
+            foreach (var id in localFunction.DescendantNodes().OfType<IdentifierNameSyntax>())
+            {
+                var symbol = semanticModel.GetSymbolInfo(id).Symbol;
+                if (symbol?.OriginalDefinition is ITypeParameterSymbol typeParameter)
+                {
+                    unusedTypeParameters.Remove(typeParameter);
+                }
+            }
+
+            typeParameters.RemoveRange(unusedTypeParameters.Except(captureTypes));
```

Comment:
there's something confusing about these three lists.  can you explain what youre' trying to do here (in comments)?

Reply:
Added comment on the caller